pkg/certificatesigningrequest/controller.go

```diff
@@ -0,0 +1,168 @@
+
+// Reconciler is the controller for CertificateSigningRequest.
+type Reconciler struct {
+	kclient        clientkubernetes.Interface
+	providerChecks ProviderChecks
+}
+
+// NewCsrController returns a new CertificateSigningRequest controller.
+func NewCsrController(kclient clientkubernetes.Interface, fn ProviderChecks) *Reconciler {
+	return &Reconciler{
+		kclient:        kclient,
+		providerChecks: fn,
+	}
+}
+
+// Run the CertificateSigningRequest controller.
+//
+//nolint:gocyclo
+func (r *Reconciler) Run(ctx context.Context) {
+	watchTimeoutSeconds := int64(time.Minute * 5)
+
+	for {
+		watcher, err := r.kclient.
+			CertificatesV1().
+			CertificateSigningRequests().
+			Watch(ctx, metav1.ListOptions{
+				Watch:          true,
+				TimeoutSeconds: &watchTimeoutSeconds, // Default timeout: 20 minutes.
+			})
+		if err != nil {
+			klog.Errorf("CertificateSigningRequestReconciler: failed to list CSR resources: %v", err)
+			time.Sleep(10 * time.Second) // Pause for a while before retrying, otherwise we'll spam error logs.
+
+			continue
+		}
+
+		csrWatcher := k8swatch.Filter(watcher, func(in k8swatch.Event) (out k8swatch.Event, keep bool) {
+			if in.Type != k8swatch.Added {
+				return in, false
+			}
+
+			return in, true
+		})
+
+	watch:
+		for {
+			select {
+			case <-ctx.Done():
+				klog.V(4).Infof("CertificateSigningRequestReconciler: context canceled, terminating")
+
+				return
+
+			case event, ok := <-csrWatcher.ResultChan():
+				if !ok {
+					// Server timeout closed the watcher channel, loop again to re-create a new one.
+					klog.V(5).Infof("CertificateSigningRequestReconciler: API server closed watcher channel")
+
+					break watch
+				}
+
+				csr, ok := event.Object.DeepCopyObject().(*certificatesv1.CertificateSigningRequest)
+				if !ok {
+					klog.Errorf("CertificateSigningRequestReconciler: expected event of type *CertificateSigningRequest, got %v",
+						event.Object.GetObjectKind())
+
+					continue
+				}
+
+				valid, err := r.Reconcile(ctx, csr)
+				if err != nil {
+					klog.Errorf("CertificateSigningRequestReconciler: failed to reconcile CSR %s: %v", csr.Name, err)
+
+					continue
+				}
+
+				_, err = r.kclient.CertificatesV1().CertificateSigningRequests().UpdateApproval(ctx, csr.Name, csr, metav1.UpdateOptions{})
+				if err != nil {
+					klog.Errorf("CertificateSigningRequestReconciler: failed to approve/deny CSR %s: %v", csr.Name, err)
+				}
+
+				if !valid {
+					klog.V(3).Infof("CertificateSigningRequestReconciler: has been denied: %s, %+v", csr.Name, err.Error())
+				} else {
+					klog.V(3).Infof("CertificateSigningRequestReconciler: has been approved: %s", csr.Name)
+				}
+			}
+		}
+	}
+}
+
+// Reconcile the CertificateSigningRequest.
+func (r *Reconciler) Reconcile(ctx context.Context, csr *certificatesv1.CertificateSigningRequest) (bool, error) {
+	switch {
+	case len(csr.Status.Conditions) > 0:
```

Comment:
is there a case when conditions > 0, but still not approved?

Reply:
Not only one CSR-approval can be in cluster. 
We do not want to check CSR which has already result of checks. Approved or Denied.

Kubernetes approvals the kubelet client csr by itself...